src/registry.rs

```diff
@@ -0,0 +1,208 @@
+    pub fn bind_cached<I, D, F, U>(
+        &mut self,
+        cx: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        name: u32,
+        f: F,
+    ) -> Result<I, BindError>
+    where
+        D: Dispatch<I, UserData = U> + 'static,
+        I: Proxy + 'static,
+        U: Send + Sync + 'static,
+        F: FnOnce() -> (u32, U),
+    {
+        match self.cached_globals.get(&name) {
+            Some(cached) => {
+                // Ensure the requested interface is the same.
+                if I::interface().name == cached.interface {
+                    // Create a new handle for the existing global.
+                    Ok(I::from_id(cx, cached.id.clone())?)
+                } else {
+                    Err(BindError::IncorrectInterface)
+                }
+            }
+
+            // First bind of a global.
+            None => {
+                let (version, udata) = f();
+                let global = self.registry.bind::<I, _>(cx, name, version, qh, udata)?;
+
+                log::debug!(target: "sctk", "Bound new cached global [{}] {} v{}", name, I::interface().name, version);
+
+                let removed = self.cached_globals.insert(
+                    name,
+                    CachedGlobal {
+                        name,
+                        _version: version,
+                        interface: I::interface().name,
+                        id: global.id(),
+                    },
+                );
+
+                assert!(removed.is_none(), "Global was cached twice?");
+
+                Ok(global)
+            }
+        }
+    }
+}
+
+pub struct RegistryDispatch<'s, D>(
+    pub &'s mut RegistryHandle,
+    pub Vec<&'s mut dyn RegistryHandler<D>>,
+);
+
+impl<D> fmt::Debug for RegistryDispatch<'_, D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RegistryDispatch").field("handle", &self.0).finish_non_exhaustive()
+    }
+}
+
+impl<D> DelegateDispatchBase<wl_registry::WlRegistry> for RegistryDispatch<'_, D> {
+    type UserData = ();
+}
+
+impl<D> DelegateDispatch<wl_registry::WlRegistry, D> for RegistryDispatch<'_, D>
+where
+    D: Dispatch<wl_registry::WlRegistry, UserData = Self::UserData>,
+{
+    fn event(
+        &mut self,
+        _: &wl_registry::WlRegistry,
+        event: wl_registry::Event,
+        _: &(),
+        cx: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        _: &mut DataInit<'_>,
+    ) {
+        match event {
+            wl_registry::Event::Global { name, interface, version } => {
+                for handler in self.1.iter_mut() {
+                    handler.new_global(cx, qh, name, &interface[..], version, self.0);
+                }
+            }
+
+            wl_registry::Event::GlobalRemove { name } => {
+                for handler in self.1.iter_mut() {
+                    handler.remove_global(cx, name);
+                }
+            }
```

Comment:
There is a possibility of global `name` reuse in the wayland protocol, as such when the `GlobalRemove` event is received, we should remove any associated global from the cache (but we don't need to destroy the wayland object itself), otherwise calls to `bind_cached` might erroneously fail.